vulnfeeds/cmd/cperepos/main.go

```diff
@@ -0,0 +1,184 @@
+		"Vendor Website", // unlikely to be a repo, but statistically significant as a description
+		"version information",
+		"product version information",
+		"product information",
+		"Product changelog",
+		"vendor product information",
+		"Changelog",
+		"Version Information",
+		"Vendor changelog",
+		"project information",
+		"product release information",
+		"product release notes",
+		"vendor product website",
+		"Product version information",
+		"vendor changelog",
+	}
+	// These repos should never be considered authoritative for a product.
+	InvalidRepos = []string{
+		"https://github.com/CVEProject/cvelist", // Heavily in Advisory URLs, sometimes shows up elsewhere
+		"https://github.com/github/cvelist",     // Fork of the above
+	}
+	CPEDictionaryFile = flag.String("cpe_dictionary", CPEDictionaryDefault, "CPE Dictionary file to parse")
+)
+
+func LoadCPEDictionary(f string) (CPEDict, error) {
+	xmlFile, err := os.Open(f)
+	if err != nil {
+		Logger.Fatalf("Failed to open %s: %v", f, err)
+	}
+
+	defer xmlFile.Close()
+
+	byteValue, _ := ioutil.ReadAll(xmlFile)
+
+	var c CPEDict
+	xml.Unmarshal(byteValue, &c)
+
+	return c, nil
+}
+
+func main() {
+	client, err := logging.NewClient(context.Background(), projectId)
+	if err != nil {
+		log.Fatalf("Failed to create client: %v", err)
+	}
+	defer client.Close()
+	Logger.GCloudLogger = client.Logger("cperepos")
+	flag.Parse()
+
+	CPEDictionary, err := LoadCPEDictionary(*CPEDictionaryFile)
+	if err != nil {
+		Logger.Fatalf("Failed to load %s: %v", *CPEDictionaryFile, err)
+	}
+	ProductToRepo := make(map[string]*string)
+	DescriptionFrequency := make(map[string]int)
+	for _, c := range CPEDictionary.CPEItems {
+		CPE, err := cves.ParseCPE(c.CPE23.Name)
+		if err != nil {
+			Logger.Infof("Failed to parse %q", c.CPE23.Name)
+			continue
+		}
+		if CPE.Part != "a" {
+			// Not interested in hardware or operating systems.
+			continue
+		}
+		if _, exists := ProductToRepo[CPE.Product]; !exists {
+			// This way every seen product will exist in the map,
+			// and ones we never determine a repo for will have a
+			// nil value.
+			ProductToRepo[CPE.Product] = nil
+		} else {
+			if ProductToRepo[CPE.Product] != nil {
+				Logger.Infof("Already have %q for %q, skipping relookup", *ProductToRepo[CPE.Product], CPE.Product)
+				continue
+			}
+		}
+		for _, r := range c.References {
+			DescriptionFrequency[r.Description] += 1
+			repo, err := cves.Repo(r.URL)
+			if err != nil {
+				Logger.Infof("Disregarding %q for %q (%s) because %v", r.URL, CPE.Product, r.Description, err)
+				continue
+			}
+			// Disregard the repos we know we don't like.
+			if slices.Contains(InvalidRepos, repo) {
+				Logger.Infof("Disliking %q for %q (%s)", repo, CPE.Product, r.Description)
+				continue
+			}
+			// Don't consider URL descriptions not explicitly allowlisted to reduce invalidity.
+			if !slices.Contains(ValidDescriptions, r.Description) {
```

Comment:
What kind of noise gets introduced because of this? 

Since we are already vetting the URL value itself through the `Repo()` call, I wonder if this could inadvertently cause us to miss valid URLs.

Reply:
I wrote this before I added the lines immediately before it, because I'd found `Advisory` references would bring in  https://github.com/CVEProject/cvelist for everything. I subsequently added the denylist, because it was still creeping in in a few places anyway.

Trialling the removal of the allowlist brought in another 1,030 product/repo combos, so I'll pull out the allowlist in favour of the denylist.

Reply:
Removed in c6b70ea